Add docs and sample code for sharing auth cookies between ASP.NET and ASP.NET Core (#28020)
dotnet/AspNetCore.Docs: aspnetcore/security/cookie-sharing.md

## Share authentication cookies between ASP.NET 4.x and ASP.NET Core apps

> ASP.NET 4.x apps that use Katana Cookie Authentication Middleware can be configured to generate authentication cookies that are compatible with the ASP.NET Core Cookie Authentication Middleware. This can be useful if a web application consists of both ASP.NET 4.x apps and ASP.NET Core apps that must share a single sign-on experience. A specific example of such a scenario is [incrementally migrating](xref:migration/inc/overview) a web app from ASP.NET to ASP.NET Core. In such scenarios, it's common for some parts of an app to be served by the original ASP.NET app while others are served by the new ASP.NET Core app. Users should only have to sign in once, though. This can be accomplished either by using the System.Web adapters' [remote authentication](xref:migration/inc/remote-authentication) feature (which will use the ASP.NET app to sign users in) or by configuring the ASP.NET app to use Katana Cookie Authentication Middleware so that auth cookies are shared with the ASP.NET Core app.

Reviewer: ```suggestion
ASP.NET 4.x apps that use Katana Cookie Authentication Middleware can be configured to generate authentication cookies that are compatible with the ASP.NET Core Cookie Authentication Middleware. This can be useful if a web application consists of both ASP.NET 4.x apps and ASP.NET Core apps that must share a single sign-on experience. A specific example of such a scenario is [incrementally migrating](xref:migration/inc/overview) a web app from ASP.NET to ASP.NET Core. In such scenarios, it's common for some parts of an app to be served by the original ASP.NET app while others are served by the new ASP.NET Core app. Users should only have to sign in once, though. This can be accomplished by either of the following approaches:

* Using the System.Web adapters' [remote authentication](xref:migration/inc/remote-authentication) feature, which uses the ASP.NET app to sign users in.
* Configuring the ASP.NET app to use Katana Cookie Authentication Middleware so that authentication cookies are shared with the ASP.NET Core app.
```

> To configure ASP.NET Katana Cookie Authentication Middleware to share cookies with an ASP.NET Core app, first follow the instructions above to configure the ASP.NET Core app to use a specific cookie name, app name, and to persist data protection keys to a well-known location (such as a shared file location).

Reviewer: First of all, first, next, finally are banned words in this repo.  Second, you can't use first, next, finally. Finally, never use finally. If they absolutely must go in that order, put then in an ordered list. Probably should put all of these in an unordered list.

[Use bigger ideas, fewer words](https://learn.microsoft.com/en-us/style-guide/top-10-tips-style-voice#use-bigger-ideas-fewer-words) : Our modern design hinges on crisp minimalism. Shorter is always better.
[Short sentences and fragments are easier to scan and read.](https://learn.microsoft.com/en-us/style-guide/brand-voice-above-all-simple-human#a-focus-on-the-customer) Prune every excess word.

```suggestion
To configure ASP.NET Katana Cookie Authentication Middleware to share cookies with an ASP.NET Core app, follow the preceding instructions to configure the ASP.NET Core app to use a specific cookie name, app name, and to persist data protection keys to a well-known location. See [Configure ASP.NET Core Data Protection](xref:security/data-protection/configuration/overview) for more information on persisting data protection keys.
```

> Next, in the ASP.NET app, install the [`Microsoft.Owin.Security.Interop`](https://www.nuget.org/packages/Microsoft.Owin.Security.Interop/) package.

Reviewer: ```suggestion
In the ASP.NET app, install the [`Microsoft.Owin.Security.Interop`](https://www.nuget.org/packages/Microsoft.Owin.Security.Interop/) package.
```

> Finally, update the `UseCookieAuthentication` call in Startup.Auth.cs to configure an AspNetTicketDataFormat to match the ASP.NET Core app's settings:

Reviewer: ```suggestion
Update the `UseCookieAuthentication` call in Startup.Auth.cs to configure an AspNetTicketDataFormat to match the ASP.NET Core app's settings:
```

[!code-csharp[](~/security/cookie-sharing/samples/WebCookieShare-NetFx/App_Start/Startup.Auth.cs?name=snippet_netfx_cookie_auth&highlight=11-23)]

Important items configured here include:

* The cookie name is set to the same name as in the ASP.NET Core app.

> * A data protection provider is created using the same key ring path. Note that in these examples, data protection keys are stored on disk but other data protection providers can be used, as well (Redis or Azure Blob Storage, for example) as long as the configuration matches between the apps.

Reviewer: ```suggestion
* A data protection provider is created using the same key ring path. Note that in these examples, data protection keys are stored on disk but other data protection providers can be used. For example, Redis or Azure Blob Storage can be used for data protection providers as long as the configuration matches between the apps. See [Configure ASP.NET Core Data Protection](xref:security/data-protection/configuration/overview) for more information on persisting data protection keys.
```

* The app name is set to be the same as the app name used in the ASP.NET Core app.
* The authentication type is set to the name of the authentication scheme in the ASP.NET Core app.

> * `System.Web.Helpers.AntiForgeryConfig.UniqueClaimTypeIdentifier` is set to a claim from the ASP.NET Core identity that will be unique to a user.

Reviewer 2: You've verified these steps still work?

Author: I verified all except the last bullet when we enabled this in our System.Web adapters samples. Those steps were sufficient to use the same cookie in both apps. 

The last bullet (UniqueClaimTypeIdentifier) I took from a response you made elsewhere about how to enable this, but I haven't verified that it works for antiforgery stuff in ASP.NET if the cookie comes from ASP.NET Core. Is there anything else I need to know to test that out or is signing the user in from ASP.NET Core and confirming that antiforgery works in ASP.NET all that this setting should affect?

Reviewer 2: That's a good start.

Author: Ok, I went back through the System.Web adapters samples and confirmed that these steps still work, including specifying a unique claim identifier for anti-forgery.

> Because the authentication type was changed to match the authentication scheme of the ASP.NET Core app, it's also necessary to update how the ASP.NET app generates new identities to use that same name. This is typically done in Models/IdentityModels.cs:

Reviewer: ```suggestion
Because the authentication type was changed to match the authentication scheme of the ASP.NET Core app, it's also necessary to update how the ASP.NET app generates new identities to use that same name. This is typically done in `Models/IdentityModels.cs`:
```

[!code-csharp[](~/security/cookie-sharing/samples/WebCookieShare-NetFx/Models/IdentityModels.cs?name=snippet_generate_identity&highlight=5-6)]

> With these changes made, the ASP.NET and ASP.NET Core apps will be able to use the same authentication cookies so that users logging in or out of one app will be reflected in the other app.

Reviewer: ```suggestion
With these changes, the ASP.NET and ASP.NET Core apps are able to use the same authentication cookies so that users signing in or out of one app are reflected in the other app.
```

## Additional resources

* <xref:host-and-deploy/web-farm>

> :::moniker-end

Reviewer: ```suggestion
* [A primer on OWIN cookie authentication middleware for the ASP.NET developer](https://brockallen.com/2013/10/24/a-primer-on-owin-cookie-authentication-middleware-for-the-asp-net-developer/)
:::moniker-end
```
